pkg/dispatcher/prometheus_test.go

```diff
@@ -0,0 +1,197 @@
+		},
+		{
+			name:         "nil map",
+			expected:     map[string]float64{},
+			path:         filepath.Join(dir, "nil-map"),
+			inputContent: "null\n",
+		},
+		{
+			name:         "file not exist",
+			expected:     map[string]float64{},
+			path:         filepath.Join(dir, "nil-map"),
+			inputContent: "other-job: 1\nsome-job: 23\n",
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if err := ioutil.WriteFile(tc.path, []byte(tc.inputContent), 0644); err != nil {
+				t.Errorf("Failed to write file: %s", tc.path)
+			}
+			if tc.name == "file not exist" {
+				if err := os.Remove(tc.path); err!=nil {
+					t.Errorf("Failed to delete file: %s", tc.path)
+				}
+			}
+
+			actual, actualError := LoadJobVolumes(tc.path)
+			if !reflect.DeepEqual(tc.expected, actual) {
+				t.Errorf("%s: actual differs from expected:\n%s", t.Name(), cmp.Diff(tc.expected, actual))
+			}
+			equalError(t, tc.expectedError, actualError)
+		})
+	}
+}
+
+func TestGetJobVolumesFromPrometheus(t *testing.T) {
+
+	now := time.Now().Unix()
+
+	testCases := []struct {
+		name             string
+		ctx              context.Context
+		queryFunc        func(ctx context.Context, query string, ts time.Time) (model.Value, v1.Warnings, error)
+		updateJobVolumes bool
+		expected         map[string]float64
+		expectedError    error
+	}{
+		{
+			name: "basic case",
+			queryFunc: func(ctx context.Context, query string, ts time.Time) (model.Value, v1.Warnings, error) {
+				vec := model.Vector([]*model.Sample{
+					{
+						Metric:    model.Metric(map[model.LabelName]model.LabelValue{model.LabelName("job_name"): model.LabelValue("pull-ci-some-test-job")}),
+						Value:     model.SampleValue(float64(23)),
+						Timestamp: model.Time(now),
+					},
+					{
+						Metric:    model.Metric(map[model.LabelName]model.LabelValue{model.LabelName("job_name"): model.LabelValue("release-openshift-ocp-installer-e2e-vsphere-upi-4.2")}),
+						Value:     model.SampleValue(float64(61.04382516525817)),
+						Timestamp: model.Time(now),
+					},
+				})
+				return vec, nil, nil
+			},
+			updateJobVolumes: true,
+			expected: map[string]float64{
+				"pull-ci-some-test-job":                               float64(23),
+				"release-openshift-ocp-installer-e2e-vsphere-upi-4.2": float64(61.04382516525817),
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual, err := GetJobVolumesFromPrometheus(tc.ctx, &prometheusAPIForTest{tc.queryFunc})
+			if !reflect.DeepEqual(tc.expected, actual) {
+				t.Errorf("%s: actual differs from expected:\n%s", t.Name(), cmp.Diff(tc.expected, actual))
+			}
+			equalError(t, tc.expectedError, err)
+		})
+	}
+}
+
+func equalError(t *testing.T, expected, actual error) {
+	if (expected == nil) != (actual == nil) {
+		t.Errorf("%s: expecting error \"%v\", got \"%v\"", t.Name(), expected, actual)
+	}
+	if expected != nil && actual != nil && expected.Error() != actual.Error() {
```

Comment:
Why is `reflect.DeepEqual()` not sufficient here?

Reply:
Updated.

Reply:
Changed it back.
Need nil check or we need to return in the previous case.

Reply:
reflect.Equal does not consider two nil errors to be equal?